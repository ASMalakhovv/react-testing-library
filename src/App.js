import {useEffect, useState} from "react";

import './App.css';

function App() {
    const [data, setData] = useState(null);
    const [isToggle, setIsToggle] = useState(false);
    const [value, setValue] = useState('');

    const handleBtnClick = () => {
      setIsToggle(prevState => !prevState);
    };

    useEffect(() => {
        setTimeout(() => {
            setData({})
        }, 100)
    }, [])

    return (
        <div className="App">
            <h1 data-testid="data-value">{value}</h1>
            {data && <div style={{ color: 'red' }}>data</div>}
            <h1>Hello world</h1>
            {isToggle && <span data-testid="toggle-span">toggle</span>}
            <button
                onClick={handleBtnClick}
                data-testid="btn-toggle"
            >
                click me
            </button>
            <input placeholder="inner text" value={value} onChange={(e) => setValue(e.target.value)}/>
        </div>
    );
}

export default App;
