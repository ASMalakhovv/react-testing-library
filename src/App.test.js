import {fireEvent, render, screen} from '@testing-library/react';
import App from './App';
import userEvent from "@testing-library/user-event";

describe('app', () => {
  test('renders learn react link', () => {
    render(<App/>);
    const helloWorldElem = screen.getByText(/hello world/i);
    const btnElem = screen.getByRole('button');
    const inputElem = screen.getByPlaceholderText(/inner text/i);
    expect(helloWorldElem).toBeInTheDocument();
    expect(btnElem).toBeInTheDocument();
    expect(inputElem).toBeInTheDocument();
    screen.debug()
    expect(inputElem).toMatchSnapshot()
  })

  test('проверка на отсутствие элемента', () => {
    render(<App/>);
    const elem = screen.queryByText(/hello2/i);
    expect(elem).not.toBeInTheDocument();
  });

  test('проверка элемента на наличие в дом async', async () => {
    render(<App/>);
    const elem = await screen.findByText(/data/i);
    expect(elem).toBeInTheDocument();
    expect(elem).toHaveStyle({color: 'red'});
  });

  test('Получение элементов через data аттрибут, импровизация событий', () => {
    render(<App/>);
    const spanElem = screen.queryByTestId("toggle-span")
    expect(spanElem).toBeNull()
    const btnElem = screen.getByTestId("btn-toggle")
    fireEvent.click(btnElem)
    expect(screen.getByTestId("toggle-span")).toBeInTheDocument()
    fireEvent.click(btnElem)
    expect(screen.queryByTestId("toggle-span")).toBeNull()
  });

  test('Тестирование инпута', () => {
    render(<App/>);
    const value = screen.getByTestId("data-value")
    expect(value).toContainHTML('')
    const input = screen.getByPlaceholderText(/inner text/i)
    expect(input).toBeInTheDocument()
    // Искусственное событие
    // fireEvent.input(input, {
    //   target: {value: '1234'}
    // })
    // Близко к пользователю, обрабатываются события нажатия клавиш и тд
    userEvent.type(input, '1234')
    expect(screen.getByTestId("data-value")).toContainHTML('1234')
  })
});
