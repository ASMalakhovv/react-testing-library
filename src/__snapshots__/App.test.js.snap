// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`app renders learn react link 1`] = `
<input
  placeholder="inner text"
  value=""
/>
`;
